src/main/scala/mesosphere/marathon/state/Container.scala

```diff
@@ -3,25 +3,26 @@ package mesosphere.marathon.state
 import com.wix.accord.dsl._
 import com.wix.accord._
 import mesosphere.marathon.api.v2.Validation._
-import org.apache.mesos.{ Protos => Mesos }
+
+import org.apache.mesos.Protos.ContainerInfo
+
 import scala.collection.immutable.Seq
 
-// TODO: trait Container and specializations?
-// Current implementation with type defaulting to DOCKER and docker to NONE makes no sense
-case class Container(
-    `type`: Mesos.ContainerInfo.Type = Mesos.ContainerInfo.Type.DOCKER,
-    volumes: Seq[Volume] = Nil,
-    docker: Option[Container.Docker] = None) {
+sealed trait Container {
+  val volumes: Seq[Volume]
 
-  import Container._
+  def docker(): Option[Container.Docker] = {
+    this match {
+      case docker: Container.Docker => Some(docker)
+      case _ => None
+    }
+  }
 
-  def portMappings: Option[Seq[Docker.PortMapping]] = {
-    import Mesos.ContainerInfo.DockerInfo.Network
+  def portMappings: Option[Seq[Container.Docker.PortMapping]] = {
```

Comment:
Should be overriden in Docker. If you want to provide a default implementation for all other Containers, use `None` here.